<!DOCTYPE html>
<html lang="en"><head>
<meta charset="utf-8"><title>CheckM - assessing the quality of genome bins</title>

<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="description" content="">
<meta name="author" content=""><!-- Le HTML5 shim, for IE6-8 support of HTML elements --><!--[if lt IE 9]>
<script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]--><!-- Le styles -->

<link href="./css/bootstrap.min.css" rel="stylesheet">
<link href="./css/bootstrap-responsive.min.css" rel="stylesheet">
<link href="./css/docs.css" rel="stylesheet"></head>
<body class="preview" data-spy="scroll" data-target=".subnav" data-offset="80">
<script src="./js/bsa.js"></script>
<div class="container"><header class="jumbotron subhead" id="overview"> </header>
<div class="row"> <img style="display: block; margin-left: auto; margin-right: auto;" src="./img/checkm.png"> </div>
<section id="theboss"><!-- Headings & Paragraph Copy -->
</section>
<div class="row">
<div class="span4">
<hr>
<h1>Work in Progress</h1>
<p>
This website is a place holder. CheckM is in active development and currently not suitable for general use. We expect to make a public release in early May.
</p>
<h1>Overview</h1>
<p>CheckM provides a set of tools for assessing the quality of genomes recovered from
isolates, single cells, or metagenomes. It provides robust estimates of genome completeness
and contamination by using collocated sets of genes that are ubiquitous and single-copy within
a phylogenetic lineage. Assessment of genome quality can also be examined using plots depicting
key genomic characteristics (e.g., GC, coding density) which highlight sequences outside the
expected distributions of a typical genome. CheckM also provides tools for identifying genome
bins that are likely candidates for merging based on marker set compatibility, similarity in
genomic characteristics, and proximity within a reference genome tree.</p>
<hr>
<h1>Get CheckM</h1>
<p>CheckM has been developed to work on a Linux system. It may
work elsewhere, but I haven't tried. If you try it somewhere
else then please let me know. I'd like to keep this list up-to-date.</p>
<h3>Use pip</h3>
<p>This is the recommended method for obtaining CheckM as it will
automaticaly install a number of dependencies. You'll need to have <a href="http://pypi.python.org/pypi">pip</a> installed
on your system. After that, open a terminal and it is as easy as
typing: </p>
<p>$ sudo pip install checkm-ACE</p>
<p>CheckM makes use of the following 3rd party programs which
must be available on the system path:</p>
<ul>
<li>hmmer &gt;= 3.0</li>
<li>prodigal &gt;= 2.6</li>
<li>pplacer &gt;= v1.1 (alpha15)</li>
</ul>
<h3>From source on github</h3>
<p>If you prefer this type of thing you can always try install
from source directly. You will need the following dependencies:</p>
<ul>
<li>numpy &gt;= 1.8.0</li>
<li>scipy &gt;= 0.9.0</li>
<li>matplotlib &gt;= 1.1.0</li>
<li>dendropy &gt;= 3.12.0</li>
<li>pysam &gt;= 0.7.4</li>
</ul>
<p>The 3rd party dependencies listed above must also be install
and placed on the system path. Clone the repo from github:</p>
<p>$ git clone git@github.com:Ecogenomics/CheckM.git</p>
<p>Then change into the CheckM directory and type:</p>
<p>$ sudo python setup.py install</p>
</div>
<div class="span4">
<hr>
<h1>Use CheckM</h1>
<p>For information on using CheckM checkout our <a href="https://github.com/Ecogenomics/CheckM/wiki">wiki</a>.</p>
<p>
Before using CheckM you need a set of putative genomes. These may come from isolates, single cells, or metagenomic data.
Our companion tool <a href="http://minillinim.github.io/GroopM/">GroopM</a> can be used to recover genomes from metagenomic
data.</p>
<h1>Cite CheckM</h1>
<p>If you find this software useful, we'd love for you to cite us:</p>
<ul>
<li>Parks DH, Imelfort M, Skennerton CT, Hugenholtz P, Tyson GW. 2014. Assessing the quality of microbial genomes recovered
from isolates, single cells, and metagenomes. In preparation.</li>
</ul>
<hr>
<h1>Talk to us</h1>
<p>All CheckM related suggestions, criticisms or abuse should be
directed to Donovan Parks:<br>
</p>
<ul>
<li>donovan_dot_parks_at_gmail_dot_com</li>
</ul>
<hr>
<h1>Licensing</h1>
<p>CheckM is licensed using the GNU General Public License
version 3 as published by the Free Software Foundation.</p>
<p>This site is maintained by Donovan Parks
and Mike Imelfort. The CheckM logo is a product of Mike Imelfort's mind.</p>
<p>This site was created using a template created by the
wonderful people at <a href="http://bootswatch.com">bootswatch</a>.
</p>
</div>
</div>
<hr><!-- Placed at the end of the document so the pages load faster -->
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
<script src="./js/bootstrap.min.js"></script> </div>
</body></html>
